### PR TITLE
get to_ldlf() working

### PR DESCRIPTION
**Pull request by francescofuggitti**
## Proposed changes

- get the `to_ldlf()` method working for all formulas except for LTLfEnd which is not supported yet.

## Types of changes

What types of changes does your code introduce to agents-tac?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../master/CONTRIBUTING.rst) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: flloat/ltlf.py [hunk 1]
```diff
@@ -535,6 +605,15 @@ def _delta(self, i: PropositionalInterpretation, epsilon=False):
         """Apply the delta function."""
         return self.to_nnf()._delta(i, epsilon=epsilon)
 
+    def to_ldlf(self):
+        """Convert the formula to LDLf."""
+        return LDLfNot(
+            LDLfDiamond(
+                RegExpStar(RegExpPropositional(PLTrue())),
+                LDLfAnd([LDLfNot(self.f.to_ldlf()), LDLfNot(LDLfEnd())]),
```
Comment (marfvr) [inline flloat/ltlf.py]: this looks wrong. It should be the dual of the Diamond formula of the eventually

Comment (francescofuggitti) [reply flloat/ltlf.py]: true, but making it an LDLfBox does not work..

Comment (marfvr) [reply flloat/ltlf.py]: namely:
```
!F(!p) = G(p)

!<true*>(!p & !end) = [true*](p | end)

```

Comment (marfvr) [reply flloat/ltlf.py]: > this looks wrong. It should be the dual of the Diamond formula of the eventually

Sorry I haven't seen the outer not. It is correct, however we could use the Box so the formula is simpler.

File: flloat/ltlf.py [hunk 2]
```diff
@@ -479,6 +531,23 @@ def _delta(self, i: PropositionalInterpretation, epsilon=False):
...
+            if len(self.formulas) > 2
+            else LDLfNot(self.formulas[1].to_ldlf())
+        )
+        return LDLfNot(
+            LDLfDiamond(
+                RegExpStar(
+                    RegExpSequence([RegExpTest(f1), RegExpPropositional(PLTrue())])
+                ),
+                LDLfAnd([f2, LDLfNot(LDLfEnd())]),
+            )
```
Comment (marfvr) [inline flloat/ltlf.py]:
```
!(!f1 Until !f2) = f1 Release f2

!<(!f1?;true)*>(!f2 & !end) = [(f1?;true)*](f2 | end)

```

Comment (marfvr) [reply flloat/ltlf.py]: Also change the Not at 540

Comment (francescofuggitti) [reply flloat/ltlf.py]: I think it should be like this:
```
!<(!f1?;true)*>(!f2 & !end) = [((!f1)?;true)*](f2 | end)
```

Comment (marfvr) [reply flloat/ltlf.py]: I edited the outer not, thanks, that was missing.

But no, I believe the not must be forwarded to `f1` in the test regex

Comment (marfvr) [reply flloat/ltlf.py]: resolved privately, the regex must not be touched by the duality of diamond and box operators.

File: flloat/ltlf.py [hunk 3]
```diff
@@ -373,8 +418,14 @@ def _delta(self, i: PropositionalInterpretation, epsilon=False):
...
-    # def to_ldlf(self):
-    #     return self.convert().to_ldlf()
+    def to_ldlf(self):
+        """Convert the formula to LDLf."""
+        return LDLfNot(
+            LDLfDiamond(
+                RegExpPropositional(PLTrue()),
+                LDLfAnd([LDLfNot(self.f.to_ldlf()), LDLfNot(LDLfEnd())]),
+            )
+        )
```
Comment (marfvr) [inline flloat/ltlf.py]:
```
!X(p) = WX(p)
!<true>(p & !end) = [true](!p | end)
```

Comment (marfvr) [reply flloat/ltlf.py]: analogous changes to the ones below

**Review by marfvr — APPROVED** [review]
Looks great! Thanks a lot :-)

File: flloat/ltlf.py (hunk 3 above)
Comment (francescofuggitti) [reply flloat/ltlf.py]: thanks. Also here: `!X(p) = WX(!p)`
so the correct transformation would be: `[true](p | end)`

**Pull request merged**